crates/cargo-bsp/src/cargo_communication/cargo_types/cargo_command.rs

```diff
@@ -1,48 +1,50 @@
-use std::path::PathBuf;
-use std::process::Command;
+use bsp_types::BuildTargetIdentifier;
+use serde_enum_str::{Deserialize_enum_str, Serialize_enum_str};
+use std::path::Path;
 
+use crate::project_model::target_details::CargoTargetKind::Lib;
+use crate::project_model::target_details::TargetDetails;
 use bsp_types::requests::{CompileParams, RunParams, TestParams};
+use std::process::Command;
 
-fn cargo_command_with_unit_graph(command_type: &str, root: PathBuf) -> Command {
-    // TODO add appropriate build target to arguments
-    let mut cmd = Command::new(toolchain::cargo());
-    cmd.current_dir(root).args([
-        "+nightly",
-        command_type,
-        "--unit-graph",
-        "-Z",
-        "unstable-options",
-    ]);
-    cmd
+#[derive(Debug, Deserialize_enum_str, Serialize_enum_str, Clone)]
+#[serde(rename_all = "camelCase")]
+enum CommandType {
+    Build,
+    Test,
+    Run,
 }
 
+const FEATURE_FLAG: &str = "--feature";
+
 pub trait CreateCommand {
     fn origin_id(&self) -> Option<String>;
 
-    fn create_unit_graph_command(&self, root: PathBuf) -> Command;
+    fn get_targets_ids(&self) -> Vec<BuildTargetIdentifier>;
+
+    fn create_unit_graph_command(&self, root: &Path, targets_details: &[TargetDetails]) -> Command;
 
-    fn create_requested_command(&self, root: PathBuf) -> Command;
+    fn create_requested_command(&self, root: &Path, targets_details: &[TargetDetails]) -> Command;
 }
 
 impl CreateCommand for CompileParams {
     fn origin_id(&self) -> Option<String> {
         self.origin_id.clone()
     }
 
-    fn create_unit_graph_command(&self, root: PathBuf) -> Command {
-        cargo_command_with_unit_graph("build", root)
+    fn get_targets_ids(&self) -> Vec<BuildTargetIdentifier> {
+        self.targets.clone()
     }
 
-    fn create_requested_command(&self, root: PathBuf) -> Command {
-        // TODO add appropriate build target to arguments
-        let mut cmd = Command::new(toolchain::cargo());
-        cmd.current_dir(root)
-            .args([
-                "build",
-                "--message-format=json",
-                self.targets[0].clone().uri.as_str(),
-            ])
-            .args(self.arguments.clone());
+    fn create_unit_graph_command(&self, root: &Path, targets_details: &[TargetDetails]) -> Command {
+        let targets_args = targets_details_to_args(targets_details);
+        cargo_command_with_unit_graph(CommandType::Build, root, targets_args)
+    }
+
+    fn create_requested_command(&self, root: &Path, targets_details: &[TargetDetails]) -> Command {
+        let targets_args = targets_details_to_args(targets_details);
+        let mut cmd = create_requested_command(CommandType::Build, root, targets_args);
+        cmd.args(self.arguments.clone());
```

Comment:
to są w ogóle zmiany z tego PRa? czy znowu coś się nie/źle zrebasowało?

Reply:
Nie były